object-store/src/main/java/org/hypertrace/config/objectstore/ScopeResolutionIdentifiedObjectStoreWithFilter.java

```diff
@@ -0,0 +1,109 @@
+import org.hypertrace.config.service.change.event.api.ConfigChangeEventGenerator;
+import org.hypertrace.config.service.v1.ConfigServiceGrpc;
+import org.hypertrace.core.grpcutils.context.RequestContext;
+
+/**
+ * ScopeResolutionIdentifiedObjectStoreWithFilter is an abstraction over {@link
+ * IdentifiedObjectStoreWithFilter} to allow scope resolution of fetched configs
+ *
+ * @param <T> Config Object
+ * @param <F> Filter
+ * @param <S> Scope
+ */
+public abstract class ScopeResolutionIdentifiedObjectStoreWithFilter<T extends Message, F, S>
+    extends IdentifiedObjectStoreWithFilter<T, F> {
+
+  protected ScopeResolutionIdentifiedObjectStoreWithFilter(
+      ConfigServiceGrpc.ConfigServiceBlockingStub configServiceBlockingStub,
+      String resourceNamespace,
+      String resourceName,
+      ConfigChangeEventGenerator configChangeEventGenerator) {
+    super(configServiceBlockingStub, resourceNamespace, resourceName, configChangeEventGenerator);
+  }
+
+  public List<T> getAllFilteredResolvedConfigs(RequestContext context, F filter, T defaultConfig) {
+    Map<S, T> configObjectsMap =
+        getAllFilteredConfigs(context, filter).stream()
+            .collect(Collectors.toMap(this::extractScope, Function.identity()));
+
+    return configObjectsMap.entrySet().stream()
+        .map(
+            entry -> resolveConfig(extractScope(entry.getValue()), defaultConfig, configObjectsMap))
+        .collect(Collectors.toUnmodifiableList());
+  }
+
+  public T getFilteredResolvedConfig(RequestContext context, F filter, S scope, T defaultConfig) {
+    List<T> filteredConfigObjects = getAllFilteredConfigs(context, filter);
+    Map<S, T> configObjectsMap =
+        filteredConfigObjects.stream()
+            .collect(Collectors.toMap(this::extractScope, Function.identity()));
+    return resolveConfig(scope, defaultConfig, configObjectsMap);
+  }
+
+  /**
+   * Method to extract scope object from given config object
+   *
+   * @param configObject
+   * @return Scope object
+   */
+  protected abstract S extractScope(T configObject);
+
+  /**
+   * Method to return scopes with which the given scope needs to be resolved in increasing order of
+   * priority For example, API scope would return a list {customer, env, service, api}
+   *
+   * @param scopeObject
+   * @return list of associated scopes in increasing order of priority
+   */
+  protected abstract List<S> getResolutionScopesWithIncreasingPriority(S scopeObject);
+
+  /**
+   * Method to merge 2 config objects Should be overridden in case the default merging logic doesn't
+   * apply, e.g. configs with arrays
+   *
+   * @param preferredConfig
+   * @param fallbackConfig
+   * @return mergedConfig
+   */
+  protected T mergeConfigs(T preferredConfig, T fallbackConfig) {
+    T.Builder builder = fallbackConfig.toBuilder();
+    mergeFromProto(builder, preferredConfig);
+    return (T) builder.build();
+  }
+
+  protected void mergeFromProto(T.Builder builder, T source) {
```

Comment:
Are we changing merge behavior such that we need to implement it custom? In other words, why can't we use the built in proto object merging, at least as the default.

Reply:
we don't use the built in proto object merging because that generally merges arrays while merging - which created issues some time back across multiple configs.. so we changed the behaviour using this custom logic to replace the arrays instead of merging..

Reply:
oh gross - is this a copy paste from proto source? If so we should call that out and explain what is changed/why.

Reply:
not exactly copy-paste but a different implementation to replace the list instead of appending - added comment explaining that

Reply:
umm not exactly.. the library code is basically `mergeFrom` for individual protos which is very different for each proto msg.. this is not exactly a copy paste but a recursive workaround (which I think @kamaleshnneerasa had got from some stackoverflow solution)